### PR TITLE
[SYCL][ESIMD] Fix a compilation of mix of unnamed ESIMD and non ESIMD kernels

### PR DESCRIPTION
**Pull request by fineg74**
The purpose of this PR is to fix an issue that leads to runtime exception when unnamed ESIMD and non ESIMD kernels are present.

### CODE REVIEW

File: llvm/tools/sycl-post-link/ModuleSplitter.cpp [hunk 1]
```diff
@@ -136,6 +136,58 @@ bool isEntryPoint(const Function &F, bool EmitOnlyKernelsAsEntryPoints) {
...
+    return make_range(AddrTakenFunctions.begin(), AddrTakenFunctions.end());
+  }
+};
+
+bool isESIMDFunctionInCallChain(const Function *F, const CallGraph &Deps) {
+  if (isESIMDFunction(*F)) {
+    return true;
+  }
+  for (const Function *F1 : Deps.successors(F)) {
+    if (isESIMDFunctionInCallChain(F1, Deps)) {
```
Comment (asudarsa) [inline llvm/tools/sycl-post-link/ModuleSplitter.cpp:185]: Is there a possibility to break out o this loop sooner?
Thanks

**Comment by kbobrovs** [comment]
@fineg74, could you please describe what the actual problem is - why unnamed ESIMD kernels lead to this problem?

**Comment by kbobrovs** [comment]
We should try to limit ESIMD specifics penetration into general-purpose tools like post-link. My suggestion would be to find the wrapper function and add 'sycl_explicit_simd' metadata to it before doing any splits in the post-link. I believe the wrapper should have `spir_kernel` calling convention, which automatically makes it an entry point. The task of finding the wrapper(s) is then searching for all `spir_kernel`s up the call chain. ESIMD splitting and lowering infra will then correctly interpret the 'sycl_explicit_simd' metadata and do what's needed.

If the wrapper does not have `spir_kernel` (and is a `spir_function`), then the user ESIMD kernel must be annotated with `spir_kernel`. In which case the search for the wrapper would be finding terminal function in the callgraph starting from the user kernel up the call chain. Callgraph traversal up the call chain is implemented in ESIMDUtils.cpp (`traverseCallgraphUp`). In this case (no `spir_kernel` in the wrapper), we should also force the wrapper to be an entry point by copying the "sycl_module_id" attribute from the user kernel or creating a dummy one.

**Comment by kbobrovs** [comment]
... I would suggest to implement finding the wrapper and adding the metadata as a pass defined within the llvm/lib/SYCLLowerIR directory and invoked in sycl-post-link.cpp somewhere around (and similarly to) the `const bool InvokeSimdMet = runModulePass<SYCLLowerInvokeSimdPass>(*M);`

**Review by asudarsa — DISMISSED** [review]
sycl-post-link change looks good. Thanks

File: llvm/lib/SYCLLowerIR/ESIMD/LowerESIMDKernelAttrs.cpp [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+using namespace llvm;
+
+namespace llvm {
+PreservedAnalyses
+SYCLLowerESIMDKernelAttrPass::run(Module &M, ModuleAnalysisManager &MAM) {
+  bool Modified = false;
+  for (Function &F : M) {
+    if (llvm::esimd::isESIMD(F)) {
+      llvm::esimd::traverseCallgraphUp(&F, [&](Function *GraphNode) {
```
Comment (kbobrovs) [inline llvm/lib/SYCLLowerIR/ESIMD/LowerESIMDKernelAttrs.cpp:27]: BTW, is the wrapper function a `spir_kernel` or `spir_function`? Is the user function  a `spir_kernel` or `spir_function` in presence of wrapper?

**Review by kbobrovs — CHANGES_REQUESTED** [review]

File: llvm/include/llvm/SYCLLowerIR/ESIMD/LowerESIMD.h [hunk 3]
```diff
@@ -77,8 +77,8 @@ class SYCLLowerESIMDKernelPropsPass
 };
 
 // Fixes ESIMD Kernel attributes for wrapper functions for ESIMD kernels
-class SYCLLowerESIMDKernelAttrPass
-    : public PassInfoMixin<SYCLLowerESIMDKernelPropsPass> {
+class SYCLFixupESIMDKernelWrapperAttrPass
```
Comment (kbobrovs) [inline llvm/include/llvm/SYCLLowerIR/ESIMD/LowerESIMD.h:80]: Not a good name either, my bad. As I noted, sycl_explicit_simd is a metadata. So please rename once again - `SYCLFixupESIMDKernelWrapperMDPass`. We need another iteration anyways.

File: llvm/lib/Passes/PassRegistry.def [hunk 4]
```diff
@@ -220,10 +228,8 @@ FUNCTION_ANALYSIS("regions", RegionInfoAnalysis())
 FUNCTION_ANALYSIS("no-op-function", NoOpFunctionAnalysis())
 FUNCTION_ANALYSIS("opt-remark-emit", OptimizationRemarkEmitterAnalysis())
 FUNCTION_ANALYSIS("scalar-evolution", ScalarEvolutionAnalysis())
-FUNCTION_ANALYSIS("should-not-run-function-passes",
-                  ShouldNotRunFunctionPassesAnalysis())
-FUNCTION_ANALYSIS("should-run-extra-vector-passes",
-                  ShouldRunExtraVectorPasses())
+FUNCTION_ANALYSIS("should-not-run-function-passes", ShouldNotRunFunctionPassesAnalysis())
```
Comment (kbobrovs) [inline llvm/lib/Passes/PassRegistry.def:231]: unrelated changes

**Review by kbobrovs — APPROVED** [review]

File: llvm/lib/SYCLLowerIR/ESIMD/LowerESIMDKernelAttrs.cpp (hunk 2 above)
Comment (kbobrovs) [reply llvm/lib/SYCLLowerIR/ESIMD/LowerESIMDKernelAttrs.cpp]: For the record: as we talked, the user function is `spir_func`, which is called by two kernels with `spir_kernel` CC - the normal and the one with the rounded range.

**Review by asudarsa — APPROVED** [review]
Changes look good. Thanks.

**Pull request merged**